### PR TITLE
ResetEvent replacing ThreadParker + std.Mutex improvements

### PR DESCRIPTION
**Pull request by kprotty**
`std.ThreadParker` and `std.Mutex` from the [previous PR](https://github.com/ziglang/zig/pull/3585) currently work and are relatively fast but there are a few improvements that can be made:

### ThreadParker
The existing implementation was fine for a simple futex but was a bit limiting as a library component:
* Didn't provide a way to do bounded blocking through `park()`
* The windows implementation had a subtle race condition for tracking waiters which would occasionally deadlock

The new `ResetEvent` addresses this by simulating Windows Event Objects instead of linux futexes and supports timeout backed waiting. The pro is that the implementation stops missed wake-ups on the windows side and does less unnecessary wake-ups on other platforms. The con is that it no longer supports multiple waiters which may be addressable in the future.

### Mutex
The new mutex improves upon the existing one in a few ways while still keeping performance essentially [on-par or better](https://github.com/kprotty/zig-adaptive-lock/tree/reset_event/results) compared to the current implementation by utilizing the word_lock found in [rust parking_lot](https://github.com/Amanieu/parking_lot/blob/master/core/src/word_lock.rs):
* `std.Mutex` is now only a word (`usize`) large which is considerably smaller than the large pthread_mutex_t and pthread_cond_t sizes on some platforms
* static initialization remains and `deinit()` should now be a no-op

The zig implementation differs slightly compared to the rust one as it unblocks threads via LIFO instead of FIFO. Its inherently unfair but it saves from traversing a linked list and there didn't seem to be such guarantees on `std.Mutex` to begin with.

### CODE REVIEW

**Comment by andrewrk** [comment]
According to the CI it looks good everywhere except macOS:

```
179/1249 reset_event.test "std-native-Debug--multi std.ResetEvent"...test failure
/Users/runner/runners/2.160.1/work/1/s/lib/std/testing.zig:149:14: 0x100392e28 in _testing.expect (test.o)
/Users/runner/runners/2.160.1/work/1/s/lib/std/reset_event.zig:400:19: 0x100416528 in _reset_event.test "std-native-Debug--multi std.ResetEvent" (test.o)
/Users/runner/runners/2.160.1/work/1/s/lib/std/special/test_runner.zig:20:25: 0x100a27754 in _std.special.main (test.o)
/Users/runner/runners/2.160.1/work/1/s/lib/std/special/start.zig:242:37: 0x100a27493 in _main (test.o)
???:?:?: 0x7fff705813d4 in ??? (???)
```

**Comment by kprotty** [comment]
Tried using std.time's approach of getting the current time but I don't have a mac to test it on so am mostly blind guessing. Looks like it may be sleeping longer (>50ms) than expected

**Comment by andrewrk** [comment]
Generally, I don't think it's sound to depend on sleeping for less than any duration of time in a test. I think that starts to get into performance testing, and that's measured separately.

Want some help with the macOS stuff? I can look into it.

**Comment by kprotty** [comment]
Sure, thanks for that

File: lib/std/reset_event.zig [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+            }
+        }
+        return true;
+    }
+};
+
+test "std.ResetEvent" {
+    // TODO
+    if (builtin.single_threaded)
+        return error.SkipZigTest;
```
Comment (andrewrk) [inline lib/std/reset_event.zig]: It looks like you did write code to handle this case, did you mean to enable this test case now?

Comment (andrewrk) [reply lib/std/reset_event.zig]: Oh, I see, the test spawns threads.

**Pull request closed without merging**